src/smt/translate_exprs.rs

```diff
@@ -682,6 +736,70 @@ impl<'smt, 'ctx> TranslateExprs<'smt, 'ctx> {
     }
 }
 
+/// Type that determines how the fuel parameter for uninterpreted functions should be synthesised.
+pub enum FuelContext {
+    /// In the head of a defining axiom a non-zero universally quantified fuel parameter is
+    /// synthesised i.e. `Succ(fuel)`.
+    Head,
+    /// In the body of a defining axiom a universally quantified fuel parameter is
+    /// synthesised i.e. `fuel`.
+    Body,
+    /// In other positions the function application represents a call. A fixed fuel
+    /// parameter is synthesized that determines how often the function definition can
+    /// be instantiated i.e. `Succ(Zero)`.
+    Call,
+}
+
+enum FuelContextInternal<'ctx> {
+    Head(QuantifiedFuel<'ctx>),
+    Body(QuantifiedFuel<'ctx>),
+    Call,
+}
+
+impl<'ctx> FuelContextInternal<'ctx> {
+    fn quantified_fuel(&self) -> Option<&QuantifiedFuel<'ctx>> {
+        match self {
+            FuelContextInternal::Head(qf) => Some(qf),
+            FuelContextInternal::Body(qf) => Some(qf),
+            FuelContextInternal::Call => None,
+        }
+    }
+
+    fn take_quantified_fuel(&mut self) -> Option<QuantifiedFuel<'ctx>> {
+        let do_take = |qf: &mut QuantifiedFuel<'ctx>| QuantifiedFuel::new(qf.0.take());
+        match self {
+            FuelContextInternal::Head(qf) => Some(do_take(qf)),
+            FuelContextInternal::Body(qf) => Some(do_take(qf)),
+            FuelContextInternal::Call => None,
+        }
+    }
+}
+
+#[derive(Default)]
```

Comment:
Add a comment saying that this is a lazily initialized fuel's `ScopeSymbolic`. Explain where it is initialized and when not.